sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -1498,14 +1500,15 @@ case class StringLocate(substr: Expression, str: Expression, start: Expression)
           } else {
             l.asInstanceOf[UTF8String].indexOf(
               r.asInstanceOf[UTF8String],
-              s.asInstanceOf[Int] - 1) + 1
+              s.asInstanceOf[Int] - 1, collationId) + 1
```

Comment:
in order to avoid performance regression for binary collation, I think it's best to short-circuit here (call `.indexOf` without collationId) using something like: `if (CollationFactory.fetchCollation(collationId).isBinaryCollation)`

for more context: https://github.com/apache/spark/pull/45216#discussion_r1499331276